### ISSUE TITLE
[Error] Cannot infer shapes or values for node "StatefulPartitionedCall/map/TensorArrayV2_2"

### ISSUE THREAD

**Issue by waghts95** [openvinotoolkit/openvino#4305]
##### System information (version)

- OpenVINO => openvino_2021.2.185
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017
- Problem classification: Model Conversion
- Framework: TensorFlow 
- Model name: efficientdet-d0

First, thank you very much openvino team for your amazing work.

I have trained efficientdet-d0 (tensorflow) model. I am trying to convert it to IR.
My command is : 
python mo_tf.py --saved_model_dir D:\rough\ed0\content\fine_tuned_model\saved_model

This is I am getting

Model Optimizer arguments:
Common parameters:
        - Path to the Input Model:      None
        - Path for generated IR:        D:\intel\openvino_2021.2.185\deployment_tools\model_optimizer\.
        - IR output name:       saved_model
        - Log level:    ERROR
        - Batch:        Not specified, inherited from the model
        - Input layers:         Not specified, inherited from the model
        - Output layers:        Not specified, inherited from the model
        - Input shapes:         Not specified, inherited from the model
        - Mean values:  Not specified
        - Scale values:         Not specified
        - Scale factor:         Not specified
        - Precision of IR:      FP32
        - Enable fusing:        True
        - Enable grouped convolutions fusing:   True
        - Move mean values to preprocess section:       None
        - Reverse input channels:       False
TensorFlow specific parameters:
        - Input model in text protobuf format:  False
        - Path to model dump for TensorBoard:   None
        - List of shared libraries with TensorFlow custom layers implementation:        None
        - Update the configuration file with input/output node names:   None
        - Use configuration file used to generate the model with Object Detection API:  None
        - Use the config file:  None
Model Optimizer version:        2021.2.0-1877-176bdf51370-releases/2021/2
2021-02-12 11:28:35.356671: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2021-02-12 11:28:35.356792: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2021-02-12 11:28:38.857191: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-02-12 11:28:38.858429: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library nvcuda.dll
2021-02-12 11:28:38.890907: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2070 computeCapability: 7.5
coreClock: 1.62GHz coreCount: 36 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.29GiB/s
2021-02-12 11:28:38.892060: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2021-02-12 11:28:38.893027: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cublas64_11.dll'; dlerror: cublas64_11.dll not found
2021-02-12 11:28:38.893970: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cublasLt64_11.dll'; dlerror: cublasLt64_11.dll not found
2021-02-12 11:28:38.896822: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2021-02-12 11:28:38.898212: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2021-02-12 11:28:38.902908: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2021-02-12 11:28:38.903999: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cusparse64_11.dll'; dlerror: cusparse64_11.dll not found
2021-02-12 11:28:38.905013: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudnn64_8.dll'; dlerror: cudnn64_8.dll not found
2021-02-12 11:28:38.905125: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1757] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2021-02-12 11:28:38.906033: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-02-12 11:28:38.906547: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-02-12 11:28:38.907215: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]
2021-02-12 11:28:38.907678: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-02-12 11:28:56.497451: I tensorflow/core/grappler/devices.cc:69] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 1
2021-02-12 11:28:56.497732: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2021-02-12 11:28:56.505751: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2070 computeCapability: 7.5
coreClock: 1.62GHz coreCount: 36 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.29GiB/s
2021-02-12 11:28:56.506884: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2021-02-12 11:28:56.507858: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cublas64_11.dll'; dlerror: cublas64_11.dll not found
2021-02-12 11:28:56.508835: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cublasLt64_11.dll'; dlerror: cublasLt64_11.dll not found
2021-02-12 11:28:56.508965: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2021-02-12 11:28:56.509298: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2021-02-12 11:28:56.510250: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2021-02-12 11:28:56.511537: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cusparse64_11.dll'; dlerror: cusparse64_11.dll not found
2021-02-12 11:28:56.512534: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudnn64_8.dll'; dlerror: cudnn64_8.dll not found
2021-02-12 11:28:56.512659: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1757] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2021-02-12 11:28:56.678367: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-02-12 11:28:56.678487: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0
2021-02-12 11:28:56.679740: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N
2021-02-12 11:28:56.680134: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-02-12 11:28:57.152306: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:928] Optimization results for grappler item: graph_to_optimize
  function_optimizer: Graph size after: 6422 nodes (5721), 14187 edges (13479), time = 232.699ms.
  function_optimizer: Graph size after: 6422 nodes (0), 14187 edges (0), time = 82.705ms.
Optimization results for grappler item: __inference_map_while_Preprocessor_ResizeToRange_cond_false_13458_17617
  function_optimizer: function_optimizer did nothing. time = 0.001ms.
  function_optimizer: function_optimizer did nothing. time = 0ms.
Optimization results for grappler item: __inference_map_while_Preprocessor_ResizeToRange_cond_true_13457_46604
  function_optimizer: function_optimizer did nothing. time = 0ms.
  function_optimizer: function_optimizer did nothing. time = 0.001ms.
Optimization results for grappler item: __inference_map_while_body_13411_46719
  function_optimizer: Graph size after: 117 nodes (0), 126 edges (0), time = 1.269ms.
  function_optimizer: Graph size after: 117 nodes (0), 126 edges (0), time = 1.302ms.
Optimization results for grappler item: __inference_map_while_cond_13410_57848
  function_optimizer: function_optimizer did nothing. time = 0ms.
  function_optimizer: function_optimizer did nothing. time = 0ms.

2021-02-12 11:29:53.239524: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-02-12 11:29:53.239644: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]
2021-02-12 11:29:53.244203: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-02-12 11:29:53.246808: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
[ ERROR ]  Cannot infer shapes or values for node "StatefulPartitionedCall/map/TensorArrayV2_2".
[ ERROR ]  Tensorflow type 21 not convertible to numpy dtype.
[ ERROR ]
[ ERROR ]  It can happen due to bug in custom shape infer function <function tf_native_tf_node_infer at 0x000002BB00584E58>.
[ ERROR ]  Or because the node inputs have incorrect values/shapes.
[ ERROR ]  Or because input shapes are incorrect (embedded to the model or passed via --input_shape).
[ ERROR ]  Run Model Optimizer with --log_level=DEBUG for more information.
[ ERROR ]  Exception occurred during running replacer "REPLACEMENT_ID" (<class 'extensions.middle.PartialInfer.PartialInfer'>): Stopped shape/value propagation at "StatefulPartitionedCall/map/TensorArrayV2_2" node.
 For more information please refer to Model Optimizer FAQ, question #38. (https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_Model_Optimizer_FAQ.html?question=38#question-38)

**Comment by avitial** [comment]
@waghts95 thanks for reaching out. Try specifying the _--input_shape_ in your MO command and see if that makes any difference, have you tried freezing the model prior to using the Model Optimizer? If possible it would be great if you can share the model files to try and see what could be the issue here.

Just FYI, as per [efficientdet-d0-tf](https://docs.openvinotoolkit.org/2021.2/omz_models_public_efficientdet_d0_tf_efficientdet_d0_tf.html) model included in [Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo) these are the MO arguments used:
```
--input_shape=[1,512,512,3] \
--input=image_arrays \
--reverse_input_channels \
--input_model=$conv_dir/efficientdet-d0_saved_model/efficientdet-d0_frozen.pb \
--transformations_config=$mo_dir/extensions/front/tf/automl_efficientdet.json
```

~Luis

**Comment by waghts95** [comment]
@avitial, thank you very much for update. 
Even specifying the --input_shape in your MO command did not help.
I have shared model files which I trained on my dataset.
https://drive.google.com/drive/folders/1bFpaK5KAZaIuDoJV143rP9HxcdYQT2Q9?usp=sharing

**Comment by avitial** [comment]
@waghts95 I am able to see the error and I think the first error is coming from `tensorflow.python.framework.errors_impl.InternalError: Tensorflow type 21 not convertible to numpy dtype`, which I'm not entirely sure how to resolve.

We have a [tutorial](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_EfficientDet_Models.html) available on how to convert EfficientDet models from [AutoML repo](https://github.com/google/automl/tree/master/efficientdet). The tutorial makes use of EfficientDet-D4, but conversion steps should be similar for EfficientDet-D0. Which implementation of EfficientDet are you using?


~Luis

**Comment by waghts95** [comment]
Thank you very much for reply. 
I tried that tutorial, even that is not working.

**Comment by avitial** [comment]
Got it, are you carefully following the steps listed? The `Can't load save_path when it is None` error is coming from freezing the model step, is that correct? Is your custom trained model based on EfficientDet implementation in tutorial? 

The model checkpoint efficientdet-d0.tar.gz referenced in the "Pretrained EfficientDet Checkpoints" has the contents below, which differ from the model files you've provided. When I attempt to freeze your model, I get a similar error as you reported in the other issue.
```
efficientdet-d0
├── checkpoint
├── d0_coco_test-dev2017.txt
├── d0_coco_val.txt
├── model.data-00000-of-00001
├── model.index
└── model.meta
```

**Comment by waghts95** [comment]
You said files are different but they works correctly when we do predictions in tensorflow. 
Shall I provide you google colab code file where these files are generated?

**Comment by waghts95** [comment]
I am using efficientdet-d0.
I am sharing colab code with you (luis.e.avitia@intel.com) that I used for training this model.
https://colab.research.google.com/drive/1wehP9A2QzLk6PgD1DUAW-zwCWTKk6N_S#scrollTo=ypWGYdPlLRUN

**Comment by waghts95** [comment]
Please reply.

**Comment by avitial** [comment]
@waghts95 sorry for the delay in my response, it looks like this implementation of EfficientDet has some SSD architecture (SSD with EfficientNet-b4 + BiFPN). That is the reason why the steps in tutorial do not really apply to your scenario. Tutorial uses EfficientDet implementation from [AutoML repo](https://github.com/google/automl/tree/master/efficientdet).

I don't believe this model architecture is currently supported and may need to figure out how to load this non-frozen model as mentioned in the [MO Developer Guide](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow.html#loading-nonfrozen-models).

~Luis

**Comment by avitial** [comment]
@waghts95 I am currently checking with dev team for additional input, I will share any details as soon as I hear back. 

BR,
~Luis

Ref. 50525

**Comment by waghts95** [comment]
@avitial,
Did you receive any information regarding error from development team ?

**Comment by lazarevevgeny** [comment]
@waghts95, take the code from https://github.com/openvinotoolkit/openvino/pull/4772 and try to convert the model using the following command line:

`
./mo.py --saved_model_dir <model_dir>/saved_model/ --transformations extensions/front/tf/ssd_support_api_v2.4.json --tensorflow_object_detection_api <model_dir>/pipeline.config --reverse_input_channels --scale 127.5 --mean_values [127.5,127.5,127.5]`

**Comment by waghts95** [comment]
Okay

**Comment by waghts95** [comment]
@lazarevevgeny,
Thank you. This worked.

**Comment by avitial** [comment]
PR has been merged to 2021.4 master branch with hash commit 522ad39, closing issue.

**Comment by shashank332** [comment]
How to take model optimizer from issue #4772
